### PR TITLE
Add support for cedar entities validation.

### PR DESCRIPTION
**Pull request by amzn-mdamine**
Issue #, if available:

Description of changes:

This PR adds support for cedar-entity-validation.

Refactor Entity json serialization similar to release/4.0.x
- Add entity validation in rust
- Add entity validation in java
- Add unit test in java and rust.

### CODE REVIEW

**Review by andrewmwells-amazon — APPROVED** [review]

**Review by shaobo-he-aws — APPROVED** [review]

File: CedarJava/src/main/java/com/cedarpolicy/AuthorizationEngine.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * Copyright 2022-2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (andrewmwells-amazon) [inline CedarJava/src/main/java/com/cedarpolicy/AuthorizationEngine.java:2]: We can change all of these to just "Copyright Cedar Contributors"

Comment (amzn-mdamine) [reply CedarJava/src/main/java/com/cedarpolicy/AuthorizationEngine.java:2]: Yes, I saw that in release/4.0.0. I think it is better than keeping track of these dates. However, I believe the current branch uses that copyright format so I wanted to keep things consistant

File: CedarJava/src/test/java/com/cedarpolicy/EntityValidationTests.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    @Test
+    public void testEntityWithUnknownAttribute() throws AuthException {
+        Entity entity = EntityValidationTests.entityGen.arbitraryEntity();
+        entity.attrs.put("test", new PrimBool(true));
+
+        EntityValidationRequest request = new EntityValidationRequest(
+                ROLE_SCHEMA, List.of(entity));
+
+
+        AuthException exception = assertThrows(AuthException.class, () -> engine.validateEntities(request));
```
Comment (amzn-mdamine) [inline CedarJava/src/test/java/com/cedarpolicy/EntityValidationTests.java:76]: I was wondering if this is the exception that should be thrown for this test.

File: CedarJava/src/test/java/com/cedarpolicy/EntityValidationTests.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        // Create a cyclic parent relationship between the entities
+        childEntity.parentsEUIDs.add(parentEntity.getEUID());
+        parentEntity.parentsEUIDs.add(childEntity.getEUID());
+
+        EntityValidationRequest request = new EntityValidationRequest(
+                ROLE_SCHEMA,
+                List.of(parentEntity, childEntity));
+
+        // Act & Assert
+        AuthException exception = assertThrows(AuthException.class, () -> engine.validateEntities(request));
```
Comment (amzn-mdamine) [inline CedarJava/src/test/java/com/cedarpolicy/EntityValidationTests.java:101]: Same question as above

Comment (andrewmwells-amazon) [reply CedarJava/src/test/java/com/cedarpolicy/EntityValidationTests.java:101]: `InternalException` seems like the best fit.

File: CedarJava/src/test/java/com/cedarpolicy/EntityValidationTests.java (hunk 2 above)
Comment (andrewmwells-amazon) [reply CedarJava/src/test/java/com/cedarpolicy/EntityValidationTests.java:76]: `InternalException` seems like the best fit.

File: CedarJavaFFI/src/interface.rs [hunk 4]
```diff
@@ -119,12 +121,42 @@ fn call_cedar(call: &str, input: &str) -> String {
...
+    schema: Value,
+    #[serde(rename = "entities")]
+    entities: Value,
+}
+
+/// public string-based JSON interface to be invoked by FFIs. Takes in a `ParseEUIDCall`, parses it and (if successful)
+/// returns a serialized `ParseEUIDOutput`
+pub fn json_validate_entities(input: &str) -> InterfaceResult {
+    match from_str::<ValidateEntityCall>(&input).map_err(|_| "Failed to parse JSON") {
+        Err(e) => InterfaceResult::fail_internally(format!("error parsing call object: {e:}")),
```
Comment (andrewmwells-amazon) [inline CedarJavaFFI/src/interface.rs:143]: I think this should be a `fail_bad_request`

File: CedarJava/src/test/java/com/cedarpolicy/EntityValidationTests.java (hunk 2 above)
Comment (khieta) [reply CedarJava/src/test/java/com/cedarpolicy/EntityValidationTests.java:76]: `BadRequestException` is fine, or you could make a custom exception, along the lines of `InvalidEUIDException`.

File: CedarJava/src/test/java/com/cedarpolicy/EntityValidationTests.java (hunk 3 above)
Comment (khieta) [reply CedarJava/src/test/java/com/cedarpolicy/EntityValidationTests.java:101]: Same comment as above

**Review by khieta — APPROVED** [review]
Changes look good 🙂 

Remember to port this to `release/3.4.x`, `release/4.0.x`, and `main` as well. The first branch should be an easy cherry-pick. The latter two will require a little more effort since  we've made changes to the code that this PR touches. I'd recommend putting up another PR for review against `release/4.0.x`.

**Pull request merged**